iceoryx2-bb/lock-free/src/spsc/queue.rs

```diff
@@ -215,8 +215,11 @@ impl<T: Copy, const CAPACITY: usize> Queue<T, CAPACITY> {
                         .as_ptr()
                 };
 
+                // prevent that `out` and `read_position` statements are reordered according to
+                // the AS-IF rule.
+                core::sync::atomic::fence(Ordering::SeqCst);
                 self.read_position
-                    .store(current_read_pos + 1, Ordering::Release);
+                    .store(current_read_pos + 1, Ordering::Relaxed);
```

Comment:
It seems you ported this before my fix in https://github.com/eclipse-iceoryx/iceoryx/pull/1397.

I think the fence also works but need to check first since I'm not that familiar with fences and they work in a slightly different way then atomics.

Reply:
Not sure about that. I'm not even sure anymore the bugfix from `#1397` is correct, although miri did not complain when I checked it the last time. Maybe we need a session with @budrus @MatthiasKillat and @hrenkap

Reply:
I think in this special case we really can use `AcqRel`, see https://www.modernescpp.com/index.php/fences-as-memory-barriers/

Only atomic load operations can be reordered to the top.

Reply:
I need to update my knowledge here. The blog post also talks about StoreLoad to have the capability of being reordered. I'm not quite sure if I still remember all of the tiny details which makes this stuff so tricky.